### PR TITLE
Optimise Keyword.delete/2 and friends

### PR DESCRIPTION
**Pull request by michalmuskala**
This optimises all functions similar to Keyword.delete/2:

  * Keyword.delete/3
  * Keyword.delete_first/2
  * Keyword.pop/3
  * Keyword.pop_lazy/3
  * Keyword.pop_first/3

There are three major changes for "delete" family:

  * use by-hand recursion instead of non-BIF :lists calls
  * in case we end up not deleting a value, use trow-catch to return the
    original list unmodified (this significantly lowers memory use)
  * build the result through body recursion, this way when we unwind the stack
    we free all the data and leave no garbage behind for the GC.

For the "pop" family:

  * use by-hand recursion instead of non-BIF :lists calls
  * do only a single pass over the list instead of two passes that previous
    fetch + delete algorithm did

In benchmarks, the new Keyword.delete/2 functions is twice as fast as the old
implementation. Since it's frequently used internally by the Keyword module (for
example in Keyword.put/3) this improves the efficiency of all those functions.

Benchmark reference: https://gist.github.com/michalmuskala/f7f38151495acf7ead131f80271359e7

### CODE REVIEW

File: lib/elixir/lib/keyword.ex [hunk 1]
```diff
@@ -449,9 +449,20 @@ defmodule Keyword do
   """
   @spec delete(t, key, value) :: t
   def delete(keywords, key, value) when is_list(keywords) and is_atom(key) do
-    :lists.filter(fn {k, v} -> k != key or v != value end, keywords)
+    do_delete(keywords, key, value, _deleted? = false)
```
Comment (whatyouhide) [inline lib/elixir/lib/keyword.ex]: We set out to not use `do_` in helper functions. Can we just use `delete/4` as the helper?

File: lib/elixir/lib/keyword.ex [hunk 2]
```diff
@@ -449,9 +449,20 @@ defmodule Keyword do
...
   end
 
+  defp do_delete([{key, value} | rest], key, value, _deleted?),
+    do: do_delete(rest, key, value, true)
+  defp do_delete([{_, _} = pair | rest], key, value, deleted?),
+    do: [pair | do_delete(rest, key, value, deleted?)]
+  defp do_delete([], _key, _value, _deleted? = true),
+    do: []
+  defp do_delete([], _key, _value, _deleted? = false),
+    do: throw(:not_deleted)
```
Comment (whatyouhide) [inline lib/elixir/lib/keyword.ex]: Is this faster than returning `nil` here and doing `delete(keywords, key, value, false) || keywords` in the original function?

File: lib/elixir/lib/keyword.ex (hunk 1 above)
Comment (lexmag) [inline lib/elixir/lib/keyword.ex]: @michalmuskala we're trying to get rid of all current occurrences of `do_` prefixing, and not to introduce new ones. Perhaps `delete_key_value` (`delete_key` for the other function) would be better name?

File: lib/elixir/lib/keyword.ex (hunk 2 above)
Comment (michalmuskala) [reply lib/elixir/lib/keyword.ex]: We can't return `nil` since we have a body-recursive function. This would only set the final tail to `nil`

Comment (whatyouhide) [reply lib/elixir/lib/keyword.ex]: Oh wow yes of course, sorry!

Comment (michalmuskala) [reply lib/elixir/lib/keyword.ex]: But we can totally do that in the `pop` case, I'll see how it compares

**Comment by michalmuskala** [comment]
I made changes renaming the functions to avoid the `do_` prefix. I also dropped the `pop` family of functions from this PR. It turned out the implementation was not optimal. I will send improvements to those functions in a separate PR.

**Comment by whatyouhide** [comment]
Thanks @michalmuskala! 💟

**Pull request merged**